### ISSUE TITLE
TReturn of Func<T, TResult> cannot be inferred in if T is value tuple.

### ISSUE THREAD

**Issue by kalaninja** [dotnet/roslyn#24781]
I'am not sure if it is actually a bug or not, but this behaviour seems strange to me. Please, consider the following code:

```csharp
public interface IDo<T>
{
    IResult<TReturn> Do<TReturn>(Func<T, TReturn> fn);
}

public interface IResult<TReturn>
{
}

public class Class1
{
    public void Test1(IDo<(int, int)> impl)
    {
        var case1 = impl.Do<int>(((int x, int y) a) => a.x * a.y); // compiles
        var case2 = impl.Do(a =>
        {
            (var x, var y) = a;
            return x * y;
        }); // compiles

        var case3 = impl.Do(((int x, int y) a) => a.x * a.y); // <-- does not compile
    }

    public void Test2(IDo<(int x, int y)> impl)
    {
        var case4 = impl.Do(a => a.x * a.y); // compiles
    }
}
```

Can anyone explain why `case3` fails with error
  

> Class1.cs(25, 27): [CS0411] The type arguments for method 'IDo<(int, int)>.Do<TReturn>(Func<(int, int), TReturn>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.

I don't understand why the compiler cannot infer the return type automatically.

**Comment by aluanhaddad** [comment]
@kalaninja this is perplexing.

Note that if you change
```C#
public void Test1(IDo<(int, int)> impl) => impl.Do(((int x, int y) a) => a.x * a.y);
```
to
```C#
public void Test1(IDo<(int x, int y)> impl) => impl.Do(((int x, int y) a) => a.x * a.y);
```
The code compiles, as does
```C#
public void Test1(IDo<(int, int)> impl) => impl.Do(((int, int) a) => a.Item1 * a.Item2);
```
It seems that the argument names must be in sync or return type inference fails. I thought tuple names were toothless so I'm very confused.

**Comment by gafter** [comment]
This is almost certainly a bug.

**Comment by jcouv** [comment]
Fix is in for 15.7
@kalaninja Thanks for reporting this!
